### DISCUSSION TITLE
Unity UI Not Displaying Firebase Data or Any Text

### DISCUSSION THREAD

**Discussion by moonweed123** [community/community#78125] [Programming Help]
### Body

Hello, I'm currently working on a Unity project where I need to display sensor data from Firebase Realtime Database on the UI. I have successfully established a connection with Firebase (as "Firebase connection successful" is displayed on the UI), but I'm unable to display the sensor data or any other text in my Unity UI components.

Here's my scenario:

I'm using Firebase Realtime Database to store sensor data.
The data is stored under the 'sensorData' path with UTC timestamps as keys.
The Unity script is supposed to fetch the most recent data every second and display it on the UI.
However, despite the successful Firebase connection, no sensor data or even the "Data not found" message is displayed on the UI. Below is the Unity script I'm using for this task. Can anyone help me identify what might be going wrong or missing in my script?

// Include the necessary Firebase and Unity namespaces
using Firebase;
using Firebase.Database;
using System.Collections;
using UnityEngine;
using UnityEngine.UI;

public class FirebaseSensorDataDisplay : MonoBehaviour
{
    // Define UI Text components for connection status and sensor data
    public Text connectionStatusText;
    public Text sensorDataText;
    private float refreshRate = 1.0f;
    private bool dataLoaded = false;
    private string sensorData = "";

    void Start()
    {
        // Initialize Firebase and start data loading coroutine
        FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task =>
        {
            if (task.Exception != null)
            {
                Debug.LogError("Firebase connection failed: " + task.Exception);
                UpdateConnectionStatus("Firebase connection failed");
                return;
            }

            UpdateConnectionStatus("Firebase connection successful");
            StartCoroutine(RefreshSensorDataEverySecond());
        });
    }

    IEnumerator RefreshSensorDataEverySecond()
    {
        while (true)
        {
            LoadLatestSensorData();
            yield return new WaitForSeconds(refreshRate);
        }
    }

    void Update()
    {
        // Update UI Text with loaded data
        if (dataLoaded)
        {
            UpdateSensorDataText(sensorData);
            dataLoaded = false;
        }
    }

    void LoadLatestSensorData()
    {
        // Fetch the latest sensor data from Firebase
        FirebaseDatabase.DefaultInstance
            .GetReference("sensorData")
            .OrderByKey()
            .LimitToLast(1)
            .GetValueAsync().ContinueWith(task =>
            {
                if (task.IsFaulted)
                {
                    Debug.LogError("Data loading failed: " + task.Exception);
                    sensorData = "Data loading failed";
                    dataLoaded = true;
                }
                else if (task.IsCompleted)
                {
                    DataSnapshot snapshot = task.Result;
                    if (snapshot.Exists && snapshot.ChildrenCount > 0)
                    {
                        // Process and store the fetched data
                        IEnumerator enumerator = snapshot.Children.GetEnumerator();
                        if (enumerator.MoveNext())
                        {
                            DataSnapshot latestData = (DataSnapshot)enumerator.Current;
                            sensorData = "Timestamp: " + latestData.Key + "\n";
                            foreach (DataSnapshot sensorItem in latestData.Children)
                            {
                                sensorData += sensorItem.Key + ": " + sensorItem.Value.ToString() + "\n";
                            }
                            dataLoaded = true;
                        }
                    }
                    else
                    {
                        sensorData = "Data not found";
                        dataLoaded = true;
                    }
                }
            });
    }

    void UpdateSensorDataText(string text)
    {
        // Update sensor data text on UI
        if (sensorDataText != null)
        {
            sensorDataText.text = text;
        }
    }

    void UpdateConnectionStatus(string status)
    {
        // Update connection status text on UI
        if (connectionStatusText != null)
        {
            connectionStatusText.text = status;
        }
    }
}

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by AliceKao0975** [comment]
Change "ContinueWith" to "ContinueWithOnMainThread"

https://forum.unity.com/threads/ui-text-ist-not-update-but-in-inspector-fixing-if-you-use-firebase.1362001/
